x/sequencer/keeper/hook_listener.go

```diff
@@ -27,8 +28,22 @@ func (hook rollappHook) BeforeUpdateState(ctx sdk.Context, seqAddr, rollappId st
 		return types.ErrNotProposer
 	}
 
-	if lastStateUpdateBySequencer {
-		return errorsmod.Wrap(hook.k.OnProposerLastBlock(ctx, proposer), "on proposer last block")
+	// if lastStateUpdateBySequencer is true, validate that the sequencer is in the middle of a rotation
+	if lastStateUpdateBySequencer && !hook.k.AwaitingLastProposerBlock(ctx, rollappId) {
+		return errorsmod.Wrap(gerrc.ErrInvalidArgument, "sequencer is not in the middle of a rotation")
+	}
+
+	return nil
+}
+
+func (hook rollappHook) AfterUpdateState(ctx sdk.Context, rollappID string, stateInfo *rollapptypes.StateInfo) error {
+	// if last block, handle proposer rotation
+	if stateInfo.Sequencer != stateInfo.NextProposer {
```

Comment:
seems like he can simply pass msg.Last = `false` and the rotation will still work.

in that case seems like `last` is redundant. 

probably cleaner to remove the `last` if we already have nextProposer and simply have it being deduced. 

not sure what's the advanteage to have both `NextProposer` and `last`.

Reply:
`last` is in the message, and the handler prepares the stateInfo with `nextSequencer` accordingly
`stateInfo.Sequencer != stateInfo.NextProposer` can happen only on `last`

just to avoid refactoring the hook interface